qbraid_qir/qasm3/visitor.py

```diff
@@ -319,19 +589,102 @@ def _visit_basic_gate_operation(self, operation: qasm3_ast.QuantumGate) -> None:
         logger.debug("Visiting basic gate operation '%s'", str(operation))
         op_name: str = operation.name.name
         op_qubits = self._get_op_bits(operation)
-        qir_func, op_qubit_count = map_qasm_op_to_pyqir_callable(op_name)
-        op_parameters = None
 
-        if len(operation.arguments) > 0:  # parametric gate
-            op_parameters = self._get_op_parameters(operation)
+        # Profile-aware qubit usage check
+        self._check_qubit_use_after_measurement(op_qubits)
+
+        # Use existing gate mapping logic but with profile awareness
+        gate_map = {
+            "h": qis.h if self._profile.restrictions.prefer_qis_over_native else qis.h,
+            "x": qis.x,
+            "y": qis.y,
+            "z": qis.z,
+            "s": qis.s,
+            "sdg": qis.s_adj,
+            "t": qis.t,
+            "tdg": qis.t_adj,
+            "cx": qis.cx,
+            "cnot": qis.cx,
+            "cz": qis.cz,
+            "ccx": qis.ccx,
+            "swap": qis.swap,
+            "rx": qis.rx,
+            "ry": qis.ry,
+            "rz": qis.rz,
+        }
 
-        for i in range(0, len(op_qubits), op_qubit_count):
-            # we apply the gate on the qubit subset linearly
-            qubit_subset = op_qubits[i : i + op_qubit_count]
-            if op_parameters is not None:
-                qir_func(self._builder, *op_parameters, *qubit_subset)
+        if op_name in gate_map:
+            qir_func = gate_map[op_name]
```

Comment:
If we are using the `BaseProfile` and we get an `x` gate, won't that use the `qis` based function when it fact it should have used `native`? Similar comment for all other gates